### PR TITLE
(BOLT-954) Support bolt apply in BoltSpec

### PR DESCRIPTION
**Pull request by donoghuc**
This commit adds the ability to use `bolt apply` in `BoltSpec`. An `apply_manifest` function has been added to `BoltSpec::Run` module to allow applying manifest code from a file or from a string with the `:execute` option.

### CODE REVIEW

File: spec/bolt_spec/run_spec.rb [hunk 1]
```diff
@@ -81,4 +83,39 @@
       expect(result['value']['kind']).to eq('bolt/run-failure')
     end
   end
+
+  describe 'apply_manifest' do
```
Comment (donoghuc) [inline spec/bolt_spec/run_spec.rb]: These tests will require a different config (there is a `facts` module on the modulepath and it tries to install puppet agent with `bolt` user). Will need to update.

File: spec/bolt_spec/run_spec.rb [hunk 2]
```diff
@@ -69,16 +69,70 @@
...
+
+  context 'with a target that has a puppet-agent installed' do
+    def run_root_config
+      { 'ssh' => root_config['ssh'] }
+    end
+
+    before(:all) do
+      result = run_task('puppet_agent::version', 'ssh', inventory: conn_inventory, config: run_root_config)
+      expect(result.first['status']).to eq('success')
+      unless result.first['result']['version']
```
Comment (donoghuc) [inline spec/bolt_spec/run_spec.rb]: I encountered a bug in the `puppet_agent::install`/`facts` tasks when running against the `rastasheep/ubuntu-sshd:16.04` container we build off of for `bolt-spec`.

When facter is not installed the `facts::bash` task determines the platform and release for 16.04 as expected with 
```
root@2b504359b3b0:/# grep "^NAME" /etc/os-release | cut -d'=' -f2 | sed "s/\"//g"
Ubuntu
root@2b504359b3b0:/# grep "^VERSION_ID" /etc/os-release | cut -d'=' -f2 | sed "s/\"//g"
16.04
```
The results of running facter are different than those using the shell code.
```
root@2b504359b3b0:/# /opt/puppetlabs/bin/facter --yaml os.name
os.name: Debian
root@2b504359b3b0:/# /opt/puppetlabs/bin/facter --yaml os.release.full
os.release.full: stretch/sid
```
This causes an error in the constructing the target download of the `.deb`. 
```
expected: http://apt.puppetlabs.com/puppet-release-xenial.deb
observed: http://apt.puppetlabs.com/puppet-release-.deb
```
Looks like this has been reported as well at https://tickets.puppetlabs.com/browse/BOLT-997

Comment (MikaelSmith) [reply spec/bolt_spec/run_spec.rb]: Returning Debian on Ubuntu sounds very wrong.

File: spec/lib/bolt_spec/conn.rb [hunk 3]
```diff
@@ -56,5 +56,35 @@ def conn_inventory
       end
       { "groups" => groups }
     end
+
+    def root_config
+      { 'modulepath' => File.join(__dir__, '../fixtures/apply'),
```
Comment (MikaelSmith) [inline spec/lib/bolt_spec/conn.rb]: We shouldn't hard-code this modulepath in the Conn module.

File: spec/bolt_spec/run_spec.rb [hunk 4]
```diff
@@ -69,16 +69,70 @@
...
+        result = run_task('puppet_agent::install', 'ssh', inventory: conn_inventory, config: run_root_config)
```
Comment (MikaelSmith) [inline spec/bolt_spec/run_spec.rb]: I'd be comfortable running these tests on a single target rather than making sure we get puppet 5 and puppet 6.

Comment (MikaelSmith) [reply spec/bolt_spec/run_spec.rb]: Maybe that's what we're doing with the `ssh` target?

File: lib/bolt_spec/run.rb [hunk 5]
```diff
@@ -49,6 +48,27 @@ def run_script(script, targets, arguments = nil, options = {}, config: nil, inve
       Bolt::Util.walk_keys(result, &:to_s)
     end
 
+    def apply_manifest(manifest, targets, execute: false, noop: false, config: nil, inventory: nil)
+      # The execute parameter is equivalent to the --execute option
+      if execute
```
Comment (MikaelSmith) [inline lib/bolt_spec/run.rb]: From a testing perspective it seems weird to design around `bolt apply ...`. If you're testing something with Bolt, it's probably a plan. But from a usable API that, it seems like a useful feature to include.

File: lib/bolt_spec/run.rb [hunk 6]
```diff
@@ -107,6 +127,23 @@ def run_script(script, targets, arguments = nil, options = {})
         targets = inventory.get_targets(targets)
         executor.run_script(targets, script, arguments, options)
       end
```
Comment (MikaelSmith) [inline lib/bolt_spec/run.rb]: Is indentation messed up here?

File: lib/bolt_spec/run.rb [hunk 7]
```diff
@@ -107,6 +127,23 @@ def run_script(script, targets, arguments = nil, options = {})
...
+
+      def apply_manifest(code, targets, filename = nil, noop = false)
+        ast = pal.parse_manifest(code, filename)
+
+        executor = Bolt::Executor.new(config.concurrency, @analytics, noop)
+        executor.start_plan(nil)
```
Comment (MikaelSmith) [inline lib/bolt_spec/run.rb]: Why is this necessary for apply but nothing else?

File: spec/lib/bolt_spec/conn.rb (hunk 3 above)
Comment (donoghuc) [reply spec/lib/bolt_spec/conn.rb]: Yeah, given that we are not going to run against puppet 5 and puppet 6 targets I don't think it makes sense to share these configs in the `Conn` class.

File: spec/bolt_spec/run_spec.rb (hunk 4 above)
Comment (donoghuc) [reply spec/bolt_spec/run_spec.rb]: Yeah, I was originally going to test against puppet 5 and puppet 6 but went with just testing against the single `ubuntu_node` `bolt-spec` image.

File: lib/bolt_spec/run.rb (hunk 5 above)
Comment (donoghuc) [reply lib/bolt_spec/run.rb]: Are you suggesting we do not implement the `execute` option? I agree that if you are not applying a manifest file the puppet code makes more sense in an apply block.

Comment (MikaelSmith) [reply lib/bolt_spec/run.rb]: I think what's here is ok. I'm not sure what the better default is though. Thinking about this plan-centrically the default should be `execute: true`. But I'm not really close to the people planning to use this.
